internal/cmd/gocdk/internal/docker/docker.go

```diff
@@ -0,0 +1,211 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package docker provides a client for interacting with the local Docker
+// daemon. This currently shells out to the Docker CLI, but could use the HTTP
+// API directly in the future.
+package docker
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"os/exec"
+	"strings"
+	"time"
+
+	"golang.org/x/xerrors"
+)
+
+// Client is a client for a Docker daemon. The zero value is a Docker client
+// that uses the current process's environment. Methods on Client are safe to
+// call concurrently from multiple goroutines.
+type Client struct {
+	env []string
+}
+
+// New creates a new client with the given environment variables.
+// Calling docker.New(nil) is the same as new(docker.Client) and uses the
+// current process's environment.
+func New(env []string) *Client {
+	return &Client{env: env}
+}
+
+// Image stores metadata about a Docker image.
+type Image struct {
+	ID         string
+	Repository string
+	Tag        string
+	Digest     string
+	CreatedAt  time.Time
+}
+
+// ListImages list images the Docker daemon has stored locally that match filterRef.
+func (c *Client) ListImages(ctx context.Context, filterRef string) ([]Image, error) {
+	args := []string{"images", "--digests", "--format={{json .}}"}
+	if filterRef != "" {
+		args = append(args, filterRef)
+	}
+	cmd := exec.CommandContext(ctx, "docker", args...)
+	cmd.Env = c.env
+	stderr := new(bytes.Buffer)
+	cmd.Stderr = stderr
+	out, err := cmd.Output()
+	if err != nil {
+		if stderr.Len() > 0 {
+			return nil, xerrors.Errorf("docker images:\n%s", stderr)
+		}
+		return nil, xerrors.Errorf("docker images: %w", err)
+	}
+	lf := []byte("\n")
+	out = bytes.TrimSuffix(out, lf)
+	var images []Image
+	for _, line := range bytes.Split(out, lf) {
+		var img struct {
+			ID         string
+			Repository string
+			Tag        string
+			Digest     string
+			CreatedAt  string
+		}
+		if err := json.Unmarshal(line, &img); err != nil {
+			return nil, xerrors.Errorf("docker images: parse output: %w", err)
+		}
+		createdAt, _ := time.Parse("2006-01-02 15:04:05 -0700 MST", img.CreatedAt)
+		images = append(images, Image{
+			ID:         cleanNone(img.ID),
+			Repository: cleanNone(img.Repository),
+			Tag:        cleanNone(img.Tag),
+			Digest:     cleanNone(img.Digest),
+			CreatedAt:  createdAt,
+		})
+	}
+	return images, nil
+}
+
+func cleanNone(s string) string {
+	if s == "<none>" {
+		return ""
+	}
+	return s
+}
+
+// RunOptions holds optional parameters for starting a Docker container.
+type RunOptions struct {
+	Args         []string
+	Env          []string
+	RemoveOnExit bool
+	Publish      []string
+}
+
+// Start starts a Docker container. It does not wait for the container to stop.
+// Start returns the container's ID on success.
+func (c *Client) Start(ctx context.Context, imageRef string, opts *RunOptions) (string, error) {
```

Comment:
`Run`? It's a bit weird for `Start` to take `RunOptions` and call `docker run`.

Reply:
Ping. Either way is fine.

Reply:
Missed this one, sorry. `Start` seems better to me because it implies the right semantics about not waiting for the container to stop. You could imagine a `Run` operation that does not set the `--detach` flag, but we don't have a need for one right now. Both of these would take in `RunOptions`.